### ISSUE TITLE
Hazelcast client unresponsive, OperationTimeoutException in the logs

### ISSUE THREAD

**Issue by agherardi** [hazelcast/hazelcast#5338]
Hi,
My application uses Hazelcast 3.4.2 on Windows. In my test environment, I have a cluster with 2 nodes - 192.168.20.22 and 192.168.20.25. After running an overnight performance/load test with no issues, when I stopped my test script this morning Hazelcast started "misbehaving". Specifically:
1. I have a client utility that connects to the cluster as an Hazelcast client and displays that status of some Hazelcast locks, queues and IMap's. The client hangs on a hazelcastInstance.getMap() operation.
2. In the server logs on 192.168.20.22, I see warnings like this:

```
WARN  [2015-05-20 15:01:38,033] com.hazelcast.spi.impl.BasicInvocation: [192.168.20.22]:5701 [vmscore_1-9b945d52-2b80-4844-a248-e57020a66cb2] [3.4.2] 'is-executing': false -> BasicInvocationFuture{invocation=BasicInvocation{ serviceName='hz:impl:mapService', op=PutOperation{PatternClientImplCache}, partitionId=9, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[192.168.20.25]:5701, backupsExpected=0, backupsCompleted=0}, response=null, done=false}
! com.hazelcast.core.OperationTimeoutException: No response for 120000 ms. Aborting invocation! BasicInvocationFuture{invocation=BasicInvocation{ serviceName='hz:impl:mapService', op=PutOperation{PatternClientImplCache}, partitionId=9, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[192.168.20.25]:5701, backupsExpected=0, backupsCompleted=0}, response=null, done=false} No response has been received!  backups-expected:0 backups-completed: 0
! at com.hazelcast.spi.impl.BasicInvocationFuture.newOperationTimeoutException(BasicInvocationFuture.java:309) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.waitForResponse(BasicInvocationFuture.java:246) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:193) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:173) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxySupport.invokeOperation(MapProxySupport.java:404) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxySupport.putInternal(MapProxySupport.java:365) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxyImpl.put(MapProxyImpl.java:102) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxyImpl.put(MapProxyImpl.java:89) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.resolveApplicationResponse(BasicInvocationFuture.java:387) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.resolveApplicationResponseOrThrowException(BasicInvocationFuture.java:320) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:194) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:173) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxySupport.invokeOperation(MapProxySupport.java:404) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxySupport.putInternal(MapProxySupport.java:365) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxyImpl.put(MapProxyImpl.java:102) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxyImpl.put(MapProxyImpl.java:89) ~[vmscore-fat.jar:1.7.7327.0]
```

3. I see similar warnings on 192.168.20.25: 

```
WARN  [2015-05-20 15:01:25,013] com.hazelcast.spi.impl.BasicInvocation: [192.168.20.25]:5701 [vmscore_1-9b945d52-2b80-4844-a248-e57020a66cb2] [3.4.2] 'is-executing': false -> BasicInvocationFuture{invocation=BasicInvocation{ serviceName='hz:impl:mapService', op=GetOperation{}, partitionId=129, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[192.168.20.25]:5701, backupsExpected=0, backupsCompleted=0}, response=null, done=false}
! com.hazelcast.core.OperationTimeoutException: No response for 120000 ms. Aborting invocation! BasicInvocationFuture{invocation=BasicInvocation{ serviceName='hz:impl:mapService', op=GetOperation{}, partitionId=129, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeout=60000, target=Address[192.168.20.25]:5701, backupsExpected=0, backupsCompleted=0}, response=null, done=false} No response has been received!  backups-expected:0 backups-completed: 0
! at com.hazelcast.spi.impl.BasicInvocationFuture.newOperationTimeoutException(BasicInvocationFuture.java:309) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.waitForResponse(BasicInvocationFuture.java:246) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:193) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.spi.impl.BasicInvocationFuture.get(BasicInvocationFuture.java:173) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxySupport.invokeOperation(MapProxySupport.java:404) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxySupport.getInternal(MapProxySupport.java:223) ~[vmscore-fat.jar:1.7.7327.0]
! at com.hazelcast.map.impl.proxy.MapProxyImpl.get(MapProxyImpl.java:84) ~[vmscore-fat.jar:1.7.7327.0]
```

Notice that both nodes think that 192.168.20.25 is the unresponsive target.

Any suggestions?

Thank you in advance.

**Comment by agherardi** [comment]
Looking at the output of jstack, I see several threads stuck in java.lang.ClassLoader.loadClass. For instance:

"hz._hzInstance_1_vmscore_1-9b945d52-2b80-4844-a248-e57020a66cb2.generic-operation.thread-0" prio=6 tid=0x000000001522d800 nid=0x1cd4 waiting on condition [0x00000000172ce000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x0000000624fe3360> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
	at java.util.concurrent.locks.LockSupport.park(Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(Unknown Source)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(Unknown Source)
	at java.util.concurrent.locks.ReentrantLock$NonfairSync.lock(Unknown Source)
	at java.util.concurrent.locks.ReentrantLock.lock(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap$Segment.scanAndLockForPut(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap$Segment.put(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap.putIfAbsent(Unknown Source)
	at java.lang.ClassLoader.getClassLoadingLock(Unknown Source)
	at java.lang.ClassLoader.loadClass(Unknown Source)
	at sun.misc.Launcher$AppClassLoader.loadClass(Unknown Source)
	at java.lang.ClassLoader.loadClass(Unknown Source)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:124)

I tried running jstack multiple times, and the same threads have the same stack trace.

**Comment by pveentjer** [comment]
Thanks for this excellent information this will help a lot figuring out the cause. I'll have a quick peek at the tryLoadClass; it is strange that a thread is parked here.

**Comment by pveentjer** [comment]
Can you run jstack with '-l' option added? This way we can see which thread is owning that lock the thread 'suffering' thread is waiting for. I find it strange that the access to the classloader is blocking 'indefinitely'

**Comment by agherardi** [comment]
Thank you. The jstack output doesn't show any threads owning 0x0000000624fe3360 - i.e., 0x0000000624fe3360 is not listed under any "Locked ownable synchronizers" section.

The only relevant information I found on the web is https://bugs.openjdk.java.net/browse/JDK-7011862 .

**Comment by pveentjer** [comment]
You are fast! Which Java version are you using? And would it be possible to run with a newer Java version, just to see if the problem disappears.

**Comment by agherardi** [comment]
Here's the output of java -version:
java version "1.7.0_51"
Java(TM) SE Runtime Environment (build 1.7.0_51-b13)
Java HotSpot(TM) 64-Bit Server VM (build 24.51-b03, mixed mode)

Yes, I'll try with the latest 1.7 update. I'll also try to run the app with a custom classloader that catches StackOverflowError's, to see if we're hitting the same issue described in https://bugs.openjdk.java.net/browse/JDK-7011862 .

**Comment by pveentjer** [comment]
How reliably can you reproduce this error?

**Comment by agherardi** [comment]
It's not super easy, but I should be able to make it happen again.

**Comment by pveentjer** [comment]
Do you have an update?

**Comment by bwzhang2011** [comment]
@agherardi, did you test and verify it in 3.5 E.A ?

**Comment by pveentjer** [comment]
I don't think his issue is related to HZ. But it appears to be a classloader getting stuck.

**Comment by agherardi** [comment]
We found a bug in our app that may be related to this. Since fixing that bug, I have NOT been able to reproduce this issue in my development environment.

Our testers need to validate the bug fix in their environment. Unfortunately, yesterday they couldn't run their tests due to networking issues - totally unrelated to Hazelcast, of course. I expect those issues to be fixed by Tuesday 5/26 (the 25th is a holiday in the US). At that time, I will post an update.

Thank you for keep checking on this issue.

**Comment by metanet** [comment]
Closing this one for now. @agherardi Please inform us when you have an update.

**Comment by veerumallavarapu** [comment]
Hi All ,

I am facing similar issue and my server is completely unresponsive.  And it is completely issue with one application server and rest of the application servers are responding without any issues.  From Thread dump I found that while retrieving the object from hazelcast is not completed and it in turn keeps all request in waiting state. Please see below 

"ajp-apr-0.0.0.0-8009-exec-170" daemon prio=6 tid=0x0000000048370800 nid=0x1054 in Object.wait() [0x0000000052aad000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.waitForResponse(BasicInvocation.java:723)
	- locked <0x0000000735eb1948> (a com.hazelcast.spi.impl.BasicInvocation$InvocationFuture)
	at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.get(BasicInvocation.java:697)
	at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.get(BasicInvocation.java:676)
	at com.hazelcast.map.proxy.MapProxySupport.invokeOperation(MapProxySupport.java:256)
	at com.hazelcast.map.proxy.MapProxySupport.getInternal(MapProxySupport.java:159)
	at com.hazelcast.map.proxy.MapProxyImpl.get(MapProxyImpl.java:53)
	at com.qumu.viewer.services.favorite.ClusteredUserFavoriteStore._findFavorite(ClusteredUserFavoriteStore.java:188)
	at com.qumu.viewer.services.favorite.ClusteredUserFavoriteStore.isFavorited(ClusteredUserFavoriteStore.java:72)
	at com.qumu.viewer.services.favorite.ClusteredUserFavoriteStore$$FastClassBySpringCGLIB$$b41d6daa.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:629)
	at com.qumu.viewer.services.favorite.ClusteredUserFavoriteStore$$EnhancerBySpringCGLIB$$25b32836.isFavorited(<generated>)
	at com.qumu.viewer.services.program.ProgramDetailsServiceImpl.populateFavoriteFlag(ProgramDetailsServiceImpl.java:915)
	at com.qumu.viewer.services.program.ProgramDetailsServiceImpl.fetchProgramDetails(ProgramDetailsServiceImpl.java:287)
	at com.qumu.viewer.services.program.ProgramDetailsServiceImpl$$FastClassBySpringCGLIB$$a19bdda4.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:700)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at com.mpi.framework.hibernate.MPIHibernateInterceptor.invoke(MPIHibernateInterceptor.java:57)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at com.mpi.framework.hibernate.MPIHibernateInterceptor.invoke(MPIHibernateInterceptor.java:57)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:633)
	at com.qumu.viewer.services.program.ProgramDetailsServiceImpl$$EnhancerBySpringCGLIB$$ff46df3f.fetchProgramDetails(<generated>)
	at com.qumu.vp.rest.program.ProgramResource.fetchProgram(ProgramResource.java:276)
	at com.qumu.vp.rest.program.ProgramResource$$FastClassBySpringCGLIB$$6e7766ed.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:700)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:96)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:260)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:94)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:633)
	at com.qumu.vp.rest.program.ProgramResource$$EnhancerBySpringCGLIB$$12fbf419.fetchProgram(<generated>)
	at com.mpi.mp.webutil.dataset.gen.DScom_qumu_vp_rest_program_ProgramResource_fetchProgram.invokeService(DScom_qumu_vp_rest_program_ProgramResource_fetchProgram.java)
	at com.mpi.mp.webutil.dataset.DatasetFactoryBean$AbstractReflectiveDataset$Loader.loadDataset(DatasetFactoryBean.java:431)
	at com.mpi.mp.webutil.dataset.DatasetInstance.getData(DatasetInstance.java:73)
	at com.qumu.vp.jsonmodel.JSONModelDirectiveFactory$JSONModelNode.render(JSONModelDirectiveFactory.java:135)
	at com.mpi.template.ste.Template.render(Template.java:18)
	at com.qumu.vp.jsonmodel.AbstractJSONModelController$JSONModelView.render(AbstractJSONModelController.java:189)
	at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1201)
	at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:986)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:933)
	at com.mpi.mp.webutil.spring.MPIDispatcherServlet.dispatchAfterFiltering(MPIDispatcherServlet.java:78)
	at com.mpi.mp.webutil.spring.MPIDispatcherServlet.doDispatch(MPIDispatcherServlet.java:61)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:851)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:953)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:844)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:620)
